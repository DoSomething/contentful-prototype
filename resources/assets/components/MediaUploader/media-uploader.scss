@import '../../scss/next-toolbox.scss';

.media-uploader {
  // width: 100%;
  background-color: #eee;
  color: $med-gray;
  // position: relative;

  // &:after {
  //   content: "";
  //   display: block;
  //   padding-bottom: 100%;

  //   @include media($medium) {
  //     padding-bottom: 30%;
  //   }

  //   @include media($large) {
  //     padding-bottom: 100%;
  //   }
  // }

  &:hover {
    background-color: darken(#eee, 5%);
  }

  label {
    align-items: center;
    cursor: pointer;
    // display: flex;
    // flex-direction: column;
    // height: 100%;
    // justify-content: center;
    // overflow: hidden;
    text-align: center;
    width: 100%;

  //   // Perfect vertical centering
  //   position: absolute;
  //   top: 50%;
  //   transform: perspective(1px) translateY(-50%);

    &::before {
      background: transparent url('../../images/plus_sign.svg');
      background-size: cover;
      content: '';
      display: block;
      height: 50px;
  //     margin: $half-spacing auto;
      width: 50px;
    }

  //   img {
  //     max-height: 100%;
  //   }
  }

  &.has-image {
    label {
      &::before {
        display: none;
      }
    }
  }

  // &.has-error {
  //   border: 1px solid $error-color;
  // }

  input {
    @include visually-hidden();
  }
}
