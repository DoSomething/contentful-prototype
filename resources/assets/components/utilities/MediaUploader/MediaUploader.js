/* global FileReader, URL, Blob */

import PropTypes from 'prop-types';
import classnames from 'classnames';
import { css } from '@emotion/core';
import React, { useState } from 'react';

import { processFile } from '../../../helpers/file';
import { report } from '../../../helpers/monitoring';
import plusSign from '../../../images/plus_sign.svg';

const MediaUploader = ({ label, onChange, hasError, media }) => {
  const [internalError, setInternalError] = useState(null);

  const readFile = file => {
    const fileReader = new FileReader();
    let blob;

    fileReader.readAsArrayBuffer(file);

    fileReader.onloadend = () => {
      try {
        blob = processFile(fileReader.result);

        onChange({
          file: blob,
          filePreviewUrl: URL.createObjectURL(blob),
        });
      } catch (error) {
        report(error);

        setInternalError(error.message);

        onChange({
          file: null,
          filePreviewUrl: null,
        });
      }
    };
  };

  const handleChange = event => {
    event.preventDefault();

    setInternalError(null);

    readFile(event.target.files[0]);
  };

  const { filePreviewUrl } = media;

  return (
    <label
      data-testid="media-uploader"
      htmlFor="media-uploader"
      className={classnames(
        'cursor-pointer block h-0 overflow-hidden relative bg-gray-200 text-gray-600 mb-3 w-full hover:bg-gray-300 focus:bg-gray-300',
        {
          'border border-solid border-red-500 shake': hasError || internalError,
        },
      )}
      css={css`
        padding-bottom: 100%;
      `}
    >
      <div
        className={classnames(
          'items-center flex h-full justify-center left-0 absolute top-0 w-full',
          {
            'bg-gray-100': filePreviewUrl,
            'flex-col': !filePreviewUrl,
          },
        )}
      >
        {filePreviewUrl ? (
          <img
            src={filePreviewUrl}
            alt="uploaded file"
            className="max-h-full"
          />
        ) : (
          <>
            <img src={plusSign} alt="plus symbol" className="mb-2" />

            <span
              className="underline font-semibold leading-normal text-center"
              css={css`
                max-width: 170px;
              `}
            >
              {label}
            </span>
            <p className="text-gray-600 pt-2 italic">{'must be <10MB'}</p>

            {internalError ? (
              <p className="text-red-300 p-3 mt-0">{internalError}</p>
            ) : null}
          </>
        )}
      </div>

      <input
        className="w-0"
        type="file"
        id="media-uploader"
        name="media-uploader"
        onChange={handleChange}
        required
        accept="image/png, image/jpeg"
      />
    </label>
  );
};

MediaUploader.propTypes = {
  hasError: PropTypes.bool,
  label: PropTypes.string,
  media: PropTypes.shape({
    file: PropTypes.instanceOf(Blob),
    filePreviewUrl: PropTypes.string,
  }),
  onChange: PropTypes.func.isRequired,
};

MediaUploader.defaultProps = {
  hasError: false,
  label: 'Upload Media',
  media: {
    file: null,
    filePreviewUrl: null,
  },
};

export default MediaUploader;
