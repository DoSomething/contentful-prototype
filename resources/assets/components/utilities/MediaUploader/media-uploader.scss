@import '../../../scss/next-toolbox.scss';

.media-uploader {
  background-color: #eee;
  color: $med-gray;
  margin-bottom: theme('spacing.3');
  width: 100%;

  &:hover,
  &:focus {
    background-color: darken(#eee, 5%);
  }

  label {
    cursor: pointer;
    display: block;
    height: 0;
    overflow: hidden;
    position: relative;
    padding-bottom: 100%;
    width: 100%;

    img {
      max-height: 100%;
    }
  }

  .media-uploader__content {
    align-items: center;
    background-color: #eee;
    display: flex;
    height: 100%;
    justify-content: center;
    left: 0;
    position: absolute;
    top: 0;
    width: 100%;

    &.media-uploader--action {
      span {
        font-weight: $weight-sbold;
        line-height: 1.3;
        max-width: 170px;
        text-align: center;
      }

      span::before {
        background: transparent url('../../../images/plus_sign.svg');
        background-size: cover;
        content: '';
        display: block;
        height: 50px;
        margin: 0 auto 12px;
        width: 50px;
      }
    }
  }

  &.has-image {
    label {
      &::before {
        display: none;
      }
    }

    .media-uploader__content {
      background-color: $off-white;
    }
  }

  &.has-error {
    border: 1px solid $error-color;
  }

  input {
    @include visually-hidden();
  }
}
