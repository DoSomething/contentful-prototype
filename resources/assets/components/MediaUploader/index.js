/* global FileReader, URL, Blob */

import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import { processFile } from '../../helpers';

import './media-uploader.scss';

class MediaUploader extends React.Component {
  constructor(props) {
    super(props);

    this.handleChange = this.handleChange.bind(this);
  }

  handleChange(event) {
    event.preventDefault();

    this.readFile(event.target.files[0]);
  }

  readFile(file) {
    const fileReader = new FileReader();
    let blob;

    fileReader.readAsArrayBuffer(file);

    fileReader.onloadend = () => {
      try {
        blob = processFile(fileReader.result);

        this.props.onChange({
          file: blob,
          filePreviewUrl: URL.createObjectURL(blob),
        });
      } catch (error) {
        // @todo: need a nice way to handle this, display message?
        console.log(error);
      }
    };
  }

  render() {
    const { hasError } = this.props;
    const { filePreviewUrl } = this.props.media;
    let content = null;

    if (filePreviewUrl) {
      content = (
        <div className="media-uploader__content media-uploader--file">
          <img src={filePreviewUrl} alt="uploaded file" />
        </div>
      );
    } else {
      content = (
        <div className="media-uploader__content media-uploader--action">
          <span>{this.props.label}</span>
        </div>
      );
    }

    return (
      <div className={classnames('media-uploader', { 'has-image': filePreviewUrl, 'has-error': hasError })}>
        <label htmlFor="media-uploader">
          {content}
          <input type="file" id="media-uploader" name="media-uploader" onChange={this.handleChange} />
        </label>
      </div>
    );
  }
}

MediaUploader.propTypes = {
  hasError: PropTypes.bool,
  label: PropTypes.string,
  media: PropTypes.shape({
    file: PropTypes.instanceOf(Blob),
    filePreviewUrl: PropTypes.string,
  }),
  onChange: PropTypes.func.isRequired,
};

MediaUploader.defaultProps = {
  hasError: false,
  label: 'Upload Media',
  media: {
    file: null,
    filePreviewUrl: null,
  },
};

export default MediaUploader;
