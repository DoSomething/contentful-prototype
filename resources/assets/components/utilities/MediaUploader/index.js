/* global FileReader, URL, Blob */

import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import { processFile } from '../../../helpers/file';

import './media-uploader.scss';

class MediaUploader extends React.Component {
  constructor(props) {
    super(props);

    this.handleChange = this.handleChange.bind(this);
  }

  handleChange(event) {
    event.preventDefault();

    this.readFile(event.target.files[0]);
  }

  readFile(file) {
    const fileReader = new FileReader();
    let blob;

    fileReader.readAsArrayBuffer(file);

    fileReader.onloadend = () => {
      try {
        blob = processFile(fileReader.result);

        this.props.onChange({
          file: blob,
          filePreviewUrl: URL.createObjectURL(blob),
        });
      } catch (error) {
        // @todo: need a nice way to handle this, display message?
        console.log(error);
      }
    };
  }

  render() {
    const { hasError } = this.props;
    const { filePreviewUrl } = this.props.media;
    let content = null;

    if (filePreviewUrl) {
      content = (
        <div className="media-uploader__content media-uploader--file">
          <img src={filePreviewUrl} alt="uploaded file" />
        </div>
      );
    } else {
      content = (
        <div className="media-uploader__content media-uploader--action underline">
          <span>{this.props.label}</span>
        </div>
      );
    }

    return (
      <div
        className={classnames('media-uploader', {
          'has-image': filePreviewUrl,
          'has-error shake': hasError,
        })}
      >
        <label htmlFor="media-uploader">
          {content}
          <input
            type="file"
            id="media-uploader"
            name="media-uploader"
            onChange={this.handleChange}
            required
          />
        </label>
      </div>
    );
  }
}

MediaUploader.propTypes = {
  hasError: PropTypes.bool,
  label: PropTypes.string,
  media: PropTypes.shape({
    file: PropTypes.instanceOf(Blob),
    filePreviewUrl: PropTypes.string,
  }),
  onChange: PropTypes.func.isRequired,
};

MediaUploader.defaultProps = {
  hasError: false,
  label: 'Upload Media',
  media: {
    file: null,
    filePreviewUrl: null,
  },
};

export default MediaUploader;
