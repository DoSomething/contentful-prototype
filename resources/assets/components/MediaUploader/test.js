import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import MediaUploader from './index';

test('MediaUploader snapshot test', () => {
  const component = shallow(<MediaUploader onChange={() => {}} />);
  const tree = shallowToJson(component);
  expect(tree).toMatchSnapshot();
});
