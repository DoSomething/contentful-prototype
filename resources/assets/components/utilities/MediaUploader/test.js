import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import MediaUploader from './MediaUploader';

test('MediaUploader snapshot test', () => {
  const component = shallow(<MediaUploader onChange={() => {}} />);
  const tree = shallowToJson(component);
  expect(tree).toMatchSnapshot();
});
