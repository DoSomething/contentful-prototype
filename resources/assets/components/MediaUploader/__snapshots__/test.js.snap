// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MediaUploader snapshot test 1`] = `
<div
  className="media-uploader"
>
  <label
    htmlFor="media-uploader"
  >
    <span>
      Upload Media
    </span>
    <input
      id="media-uploader"
      name="media-uploader"
      onChange={[Function]}
      type="file"
    />
  </label>
</div>
`;
